compiler/src/iree/compiler/Codegen/LLVMGPU/ROCDLBufferInstructionsOptimization.cpp

```diff
@@ -43,92 +42,86 @@ Value createI1And(Location loc, ArrayRef<Value> values, OpBuilder &builder) {
-        }
-      } else {
-        // The non-constant dimensions are only allowed to be unit to ensure
-        // an all ones or an all zeros mask.
-        if (maskShape[idx] != 1) {
-          return rewriter.notifyMatchFailure(
-              maskOp, "non-unit, non-constant dimension found");
-        }
-        ValuesToAnd.push_back(maskIndex);
+    } else {
+      if (maskShape[idx] != 1) {
+        isValid = false;
+        break;
       }
+      ValuesToAnd.push_back(maskIndex);
     }
-    // There are no non-constant unit dimensions which means this is
-    // a always one mask, such masks should be folded by vector
-    // canonicalizations so we dont handle it.
-    if (ValuesToAnd.empty()) {
-      return rewriter.notifyMatchFailure(maskOp, "trivial mask not supported");
+  }
+  // Bail out if the mask doesnt meet the criteria or
+  // is statically all 1's in which case we dont need
+  // to do anything.
+  if (!isValid || ValuesToAnd.empty()) {
+    return;
+  }
+
+  for (Operation *user : maskOp.getResult().getUsers()) {
+    auto readOp = dyn_cast<vector::TransferReadOp>(user);
+    // Only TransferReadOps are supported.
+    if (!readOp)
+      continue;
+
+    auto sourceType = dyn_cast<MemRefType>(readOp.getSource().getType());
+    // only supported for fat raw buffers.
+    if (!sourceType || !hasAMDGPUFatRawBufferAddressSpace(sourceType))
+      continue;
+
+    SmallVector<bool> inBounds = readOp.getInBoundsValues();
+    // Only supported for reads that are fully in_bounds.
+    if (inBounds.size() != sourceType.getRank() ||
+        llvm::any_of(inBounds, [](bool inBound) { return !inBound; })) {
+      continue;
     }
+
+    rewriter.setInsertionPoint(readOp);
     Value selectValue = createI1And(loc, ValuesToAnd, rewriter);
     auto constantValue = rewriter.create<vector::SplatOp>(
         loc, readOp.getVectorType(), readOp.getPadding());
 
     auto newReadOp = rewriter.create<vector::TransferReadOp>(
         loc, readOp.getVectorType(), readOp.getSource(), readOp.getIndices(),
         readOp.getPadding(), ArrayRef<bool>{inBounds});
-
     auto selectOp = rewriter.create<arith::SelectOp>(loc, selectValue,
                                                      newReadOp, constantValue);
     rewriter.replaceAllUsesWith(readOp, selectOp);
-
-    return success();
   }
-};
+  return;
+}
 
 struct ROCDLBufferInstructionsOptimizationPass final
     : impl::ROCDLBufferInstructionsOptimizationPassBase<
           ROCDLBufferInstructionsOptimizationPass> {
   void runOnOperation() override {
-    RewritePatternSet patterns(&getContext());
-    patterns.add<SimplifyMaskVectorTransferRead>(&getContext());
-    walkAndApplyPatterns(getOperation(), std::move(patterns));
+    MLIRContext *context = &getContext();
+    FunctionOpInterface funcOp = getOperation();
+    SmallVector<vector::CreateMaskOp> maskOps(
+        funcOp.getFunctionBody().getOps<vector::CreateMaskOp>());
```

Comment:
I think this can be accomplished with a WalkPatternRewriter - I forget the exact terminology, I just know it's new - instead of abandoning the pattern framework entirely

Reply:
The reason I switched is this is actually faster, the `WalkPatternRewriter` is just a walk on a funcOp, which is prone to crashing if you modify/erase the ops that are in your queue to walk. Here is the comment (and corresponding PR) that gave me the idea to do it this way. https://github.com/iree-org/iree/pull/20627#discussion_r2067452732

Reply:
Got it

Reply:
ok so I missed why that other way was fast, because it doesn't look inside the regions of ops in a nested way, since our masks/loads are inside loop we need to do that. However, using `WalkPatternRewriter` is essentially walking and changing IR at the same time and that was crashing for me, so I have gone with the current approach of doing a walk first and then calling the rewriter function . The alternative is to use greedyrewriter which is equipped to handle such case but I think we didnt want to use it for such non iterative patterns.